snuba/utils/streams/consumer.py

```diff
@@ -97,6 +97,30 @@ def unsubscribe(self) -> None:
     def poll(self, timeout: Optional[float] = None) -> Optional[Message[TPayload]]:
         raise NotImplementedError
 
+    @abstractmethod
+    def pause(self, partitions: Sequence[Partition]) -> None:
```

Comment:
What's the rationale for pausing some partitions and not others ? Also if we have pause and resume that takes partitions as parameter. Should we also have a method that gives the status of a given partitions (whether it is paused or running) ?

Reply:
> What's the rationale for pausing some partitions and not others ? 

Each partition can be thought of an independent stream at some level, so it makes sense to be able to manage them independently, the same way each partition can be `seek`ed on or have its offsets committed without taking the same action on other partitions.

For a less abstract reason, consider the synchronized consumer: offsets for each partition are committed independently, meaning some local offsets may lag behind the remote offsets for some set of partitions (so the local consumer should be allowed to consume from those partitions; in other words, resumed) while the local offsets may be equal to (or ahead of) the remote offsets for some other set of partitions (so the local consumer should be not be allowed to consume from those partitions; in other words, paused.)

> Should we also have a method that gives the status of a given partitions (whether it is paused or running) ?

I was on the fence about this, I conservatively left it out because confluent-kafka does not expose it, so there's actually no way to authoritatively say whether or not the Confluent consumer thinks a partition is paused or not. That said, if our state is different than the Confluent consumer, something is probably going poorly beyond the method returning incorrect results, so maybe it's not a big deal. It'd make the tests a little cleaner — we'd at least be able to say that we think the pause/resume _should_ have had an effect. I guess I don't have a strong opinion one way or another, happy to add it if you think it's useful, happy to leave it out if you don't see a strong case for it.

Reply:
> It'd make the tests a little cleaner — we'd at least be able to say that we think the pause/resume should have had an effect.

More thinking on this — this will likely make testing the synchronized consumer substantially easier.

Reply:
>I guess I don't have a strong opinion one way or another, happy to add it if you think it's useful, happy to leave it out if you don't see a strong case for it.

I guess it depends on how we plan to use this. Knowing that confluent does not provide us a reliable source of truth and that a partition can be transparently restarted during reassignment, I think we can add it when/if needed.